File: lru.java
import java.util.*;

public class LRUPageReplacement {

    public static void lruPageReplacement(int[] pages, int frameSize) {
        List<Integer> frames = new ArrayList<>();
        Map<Integer, Integer> recentUsage = new HashMap<>(); // Tracks last usage index
        int pageFaults = 0;

        for (int i = 0; i < pages.length; i++) {
            int page = pages[i];

            if (!frames.contains(page)) {
                pageFaults++;
                if (frames.size() < frameSize) {
                    frames.add(page);
                } else {
                    // Find the least recently used page in frames
                    int lruPage = frames.get(0);
                    int minUsage = Integer.MAX_VALUE;

                    for (int f : frames) {
                        int lastUsed = recentUsage.getOrDefault(f, -1);
                        if (lastUsed < minUsage) {
                            minUsage = lastUsed;
                            lruPage = f;
                        }
                    }

                    // Replace LRU page
                    int lruIndex = frames.indexOf(lruPage);
                    frames.set(lruIndex, page);
                }
            }

            // Update last usage of current page
            recentUsage.put(page, i);

            System.out.println("Frames after accessing " + page + ": " + frames);
        }

        System.out.println("\nTotal Page Faults: " + pageFaults);
    }

    public static void main(String[] args) {
        int[] pages = {2, 3, 2, 1, 5, 2, 4, 5, 3, 2, 5, 2};
        int frameSize = 3;

        lruPageReplacement(pages, frameSize);
    }
}

File: Python/lru.py
# Corrected LRU Page Replacement Algorithm in Python

def lru_page_replacement(pages, frame_size):
    frames = []
    page_faults = 0
    recent_usage = {}  # Tracks last usage of pages currently in frames

    for i, page in enumerate(pages):
        if page not in frames:
            page_faults += 1
            if len(frames) < frame_size:
                frames.append(page)
            else:
                # Consider only pages currently in frames for LRU
                lru_page = min({p: recent_usage[p] for p in frames}, key=lambda k: recent_usage[k])
                lru_index = frames.index(lru_page)
                frames[lru_index] = page
        # Update the last usage of the current page
        recent_usage[page] = i
        print(f"Frames after accessing {page}: {frames}")

    print(f"\nTotal Page Faults: {page_faults}")

# Sample input
pages = [2, 3, 2, 1, 5, 2, 4, 5, 3, 2, 5, 2]
frame_size = 3

lru_page_replacement(pages, frame_size)
